text/0126-introduce-xcq.md

```diff
@@ -0,0 +1,384 @@
+    method: &str,
+    input: &[u8],
+    weight_limit: u64,
+) -> Result<Vec<u8>, XcqExecutorError> {...}
+```
+
+### XCQ Extension
+
+An extension-based design is essential for several reasons:
+- Different chains may have different data types for semantically similar queries, making it challenging to standardize function calls across them.
+An extension-based design with optional associated types allows these diverse data types to be specified and utilized effectively.
+- Function calls distributed across various pallets can be amalgamated into a single extension, simplifying the development process and ensuring a more cohesive and maintainable codebase.
+- New functionalities can be added without upgrading the core part of the XCQ.
+- Ensure the core part is in a minimal scope.
+
+Essential components of an XCQ extension system include:
+
+- A hash-based extension id generation mechanism for addressing and versioning . The hash value derives from the extension name and its method sets. Any update to an extension is treated as a new extension.
+
+- `decl_extension` macro: Defines an extension as a Rust trait with optional associated types.
+
+**Example usage**:
+```rust
+use xcq_extension::decl_extension;
+
+pub trait Config {
+    type AssetId: Codec;
+    type AccountId: Codec;
+    type Balance: Codec;
+}
+decl_extension! {
+    pub trait ExtensionFungibles {
+        type Config: Config;
+        fn total_supply(asset: <Self::Config as Config>::AssetId) -> <Self::Config as Config>::Balance;
+        fn balance(asset: <Self::Config as Config>::AssetId, who: <Self::Config as Config>::AccountId) -> <Self::Config as Config>::Balance;
+    }
+}
+```
+
+- `impl_extensions` macro: Generates extension implementations and extension-level metadata.
+
+**Example Usage**:
+
+```rust
+// ExtensionImpl is an aggregate struct to impl different extensions
+struct ExtensionImpl;
+impl extension_fungibles::Config for ExtensionImpl {
+    type AssetId = u32;
+    type AccountId = [u8; 32];
+    type Balance = u64;
+}
+impl_extensions! {
+    impl extension_core::ExtensionCore for ExtensionImpl {
+        type Config = ExtensionImpl;
+        fn has_extension(id: <Self::Config as extension_core::Config>::ExtensionId) -> bool {
+            matches!(id, 0 | 1)
+        }
+    }
+
+    impl extension_fungibles::ExtensionFungibles for ExtensionImpl {
+        type Config = ExtensionImpl;
+        #[allow(unused_variables)]
+        fn total_supply(asset: <Self::Config as extension_fungibles::Config>::AssetId) -> <Self::Config as extension_fungibles::Config>::Balance {
+            200
+        }
+        #[allow(unused_variables)]
+        fn balance(asset: <Self::Config as extension_fungibles::Config>::AssetId, who: <Self::Config as extension_fungibles::Config>::AccountId) -> <Self::Config as extension_fungibles::Config>::Balance {
+            100
+        }
+    }
+}
+```
+
+- `ExtensionExecutor`: Connects extension implementations and `xcq-executor`.
```

Comment:
If one wants to expose pallet view functions or runtime APIs to the XCQ programs, do extensions have to be written which call into them (and then those extensions are what is exposed in the XCQ program)?

Is the reason for having this new extension stuff instead of exposing raw runtime APIs and pallet view functions to XCQ programs that extensions have associated types which can be tweaked by different runtimes, whereas runtime APIs do not? (If that were the only reason, I wonder whether runtime APIs could be extended to optionally support associated types to achieve the same without a new thing being needed?)

Reply:
A1: Yes, the fundamental APIs exposed to XCQ programs are organized at the extension level.

A2: It's not the only reason. Runtime APIs cannot be recognized via some identifications, while extensions can be recognized via extension ids.

Reply:
> Runtime APIs cannot be recognized via some identifications

Runtime API methods can all be identified as `$trait_name ++ _ ++ $method_name`, and so we could expose functions with similar names to the PolkaVM program? I suspect I am missing something :)

If Runtime APIs "just worked" natively, it would mean that we could provide a single stable interface to users (who can call the APIs themselves) and XCQ programs (which could call the same APIs), and wouldn't need to add another concept ro Runtimes.

Reply:
The main reason we created XCQ extension is to have content based address. i.e. The extension ID is the hash of the extension definition, which includes the methods and argument types. So we can for sure know the compatibility with a matching extension ID. Runtime API is using versioning for that, which requires extra collaboration.
Besides, Runtime API are designed to be invoked from outside of wasm, which is only half of the use case of XCQ, the other half is XCM. We cannot assume every runtime API are safe to be dispatched by XCM.

Reply:
Thankyou for your reply!

>  The extension ID is the hash of the extension definition, which includes the methods and argument types. So we can for sure know the compatibility with a matching extension ID. Runtime API is using versioning for that, which requires extra collaboration.

How does having a content based address help when compared with non content based addresses (ie the address just being the extension trait  and method name, like Runtime APIs)? 

Looking at some of the possibilities:
- **If I change the semantics but not the type information of some method:** 
   - **Using content based address:** the method will no longer do what is expected, but will still be callable via the same hash. 
   - **Using non-content based address:** same as above
   - **Verdict:** Neither approach is any different here; we require social agreement (mostly a no brainer ofc) to not change these things.
- **If I change the semantics _and_ type information (ie name, input, return types):**
   - ** Using content based address:** calling the "old" address will now fail with some error that it doesn't exist.
   - ** Using non-content based address:** calling the "old" address will still work, but fail with some type error on encoding or decoding.
   - **Verdict:** Both approaches lead to an error; we require social agreement to not change/remove methods if we want to avoid breakage.

In other words, however we approach addressing things, we still need people to agree not to make breaking changes to these functions if we want to avoid breaking compatibility. 

Do content based addresses help in some other way that I am missing (definitely likely!!)? Is the thinking here mainly around XCM ie making it more likely for queries that they call to no longer be accessible?

And a couple more Dumb Qs: what do extensions provide on top of runtime APIs that ensures that they are safe to be dispatched by XCM (or marks them as not being safe)? Might be there some extensions that we want to be able to call in XCQ programs but not via XCM?

Reply:
The current polkadot-sdk runtime calls are name based and have many known issues and most of them can be addressed by content based address, which is more commonly used. e.g. EVM selector.

Some advantages:
- The extension ID fixed the name and types, the name based approach only fixed the name
- The extension ID is hash of the whole interface, whereas the name, well, is just whatever the name referring to.
  - e.g. `0x123456` could mean `trait Fungibles { fn balanceOf(who: AccountId) -> U128; fn totalSupply() -> U128 }` but with name, `Fungibles` means a trait named Fungibles but does not constraint the methods. A name of `Fungibles.balanceOf` is just a function name, and not telling anything about other members.
- Unexpected change of type can be detected easily because it changes the extension ID. Not the case for name based ID.
- There is no need to have central name registry. 3 different chains may all want to propose their version of `Fungibles` interface and who decide who got it? Not an issue for hash.

Reply:
`PermissionController` defined in the RFC is the mechanism to only allow some extension in some context.